src/device/router_device_worker.erl

```diff
@@ -733,18 +732,45 @@ handle_cast(
                     end,
                 {devaddr, DevAddr0} = blockchain_helium_packet_v1:routing_info(Packet0),
                 DevAddr1 = lorawan_utils:reverse(<<DevAddr0:32/integer-unsigned-big>>),
+                Keys = router_device:keys(Device0),
                 DeviceUpdates = [
                     {keys,
                         lists:filter(
                             fun({NwkSKey, _}) -> NwkSKey == UsedNwkSKey end,
-                            router_device:keys(Device0)
+                            Keys
                         )},
                     {dev_nonces, DevNonces},
                     {devaddrs, [DevAddr1]},
                     {fcnt, PacketFCnt}
                 ],
                 D1 = router_device:update(DeviceUpdates, Device0),
                 ok = save_device(DB, CF, D1),
+
+                ToAdd = [{add, DevAddr0, UsedNwkSKey}],
+                %% We have to usort just in case DevAddr assigned is the same
+                ToRemove0 = lists:usort(
```

Comment:
Some alternatives.
You can `lists:flatmap/2` on the outside, and `lists:map/2` on the inside to not have to worry about the `Acc`s.

Or somehow grab all the devaddrs as ints and do a list comprehension.
```erlang
[{remove, DevAddr, Key} || DevAddr <- DevAddrInts, Key <- Keys]
```